src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,233 @@
+import {Rescuable} from 'solidity-utils/contracts/utils/Rescuable.sol';
+import {RescuableBase, IRescuableBase} from 'solidity-utils/contracts/utils/RescuableBase.sol';
+
+import {IBalancerVault} from './balancer-v2/IBalancerVault.sol';
+import {WeightedPoolUserData} from './balancer-v2/WeightedPoolUserData.sol';
+import {IBalancerStrategyManager} from './IBalancerStrategyManager.sol';
+
+/**
+ * @title Balancer V2 Weighted Pool Strategy Manager
+ * @author TokenLogic
+ * @notice A contract to manage deposits and withdrawals into a Balancer V2 Weighted Pool.
+ * @dev This contract uses the Balancer Vault to manage liquidity, using exact token amounts for BPT.
+ */
+contract BalancerV2WeightedPoolStrategyManager is
+  IBalancerStrategyManager,
+  OwnableWithGuardian,
+  Rescuable
+{
+  using SafeERC20 for IERC20;
+
+  /// @notice The Balancer Vault contract
+  IBalancerVault public immutable VAULT;
+
+  /// @notice The number of tokens in the pool
+
+  /// @notice The address of the Hypernative service
+  address public immutable HYPERNATIVE;
+
+  /// @dev Mapping of token configurations (token to provider)
+  mapping(address token => address provider) public override tokenProvider;
+
+  /// @dev Restricts access to only the owner, guardian, or Hypernative address.
+  modifier onlyWithdrawable() {
+    if (_msgSender() != owner() && _msgSender() != guardian() && _msgSender() != HYPERNATIVE) {
+      revert Unauthorized();
+    }
+    _;
+  }
+
+  /**
+   * @notice Constructor to initialize the contract.
+   * @param _vault The address of the Balancer Vault.
+   * @param _tokenConfig The array of token configurations (address and provider).
+   * @param _owner The owner of the contract.
+   * @param _guardian The guardian of the contract.
+   * @param _hypernative The Hypernative service address.
+   */
+  constructor(
+    address _vault,
+    TokenConfig[] memory _tokenConfig,
+    address _owner,
+    address _guardian,
+    address _hypernative
+  ) {
+    VAULT = IBalancerVault(_vault);
+
+    for (uint256 i = 0; i < _tokenConfig.length; ) {
+      tokenProvider[_tokenConfig[i].token] = _tokenConfig[i].provider;
+
+      emit TokenProviderUpdated(_tokenConfig[i].token, address(0), _tokenConfig[i].provider);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    _transferOwnership(_owner);
+    _updateGuardian(_guardian);
+    HYPERNATIVE = _hypernative;
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function setTokenProvider(address _token, address _provider) external onlyOwner {
+    address oldProvider = tokenProvider[_token];
+    tokenProvider[_token] = _provider;
+
+    emit TokenProviderUpdated(_token, oldProvider, _provider);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function deposit(bytes32 _poolId, uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    address[] memory assets;
+    (assets, , ) = VAULT.getPoolTokens(_poolId);
```

Comment:
What happens if poolId is invalid? Any way to check and revert that's more explicit or will VAULT revert?

Reply:
yeah. maybe. and we(owner or guardian) calls this function. so I think poolId check is unnecessary. actually I'm not able to find such function on BalacerVault interface

Reply:
@LucasWongC any updates on what happens if you give it a bad pool id? can you add a test for it?